orttraining/orttraining/python/training/onnxblock/model.py

```diff
@@ -0,0 +1,134 @@
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License.
+# model.py
+
+from abc import abstractmethod
+
+import onnx
+
+import onnxruntime.training.onnxblock._graph_utils as graph_utils
+import onnxruntime.training.onnxblock.building_blocks as building_blocks
+import onnxruntime.training.onnxblock.model_accessor as accessor
+
+
+class Model(building_blocks.Block):
+    """Builds the forward model based on user's build method."""
+
+    def __init__(self):
+        ...
+
+    @abstractmethod
+    def build(self, *args, **kwargs):
+        """Customize and build the forward graph for this model.
+
+        This method is to be overridden by the user's implementation.
+        """
+        ...
+
+    def __call__(self, *args, **kwargs):
+        """Calls the user's build method and runs validation on top.
+
+        The output onnx model is got by invoking the user's build method.
+        """
+        # build the user model
+        output = self.build(*args, **kwargs)
+
+        # Perform shape inference
+        model_with_shapes = onnx.shape_inference.infer_shapes(accessor.global_accessor.model)
+        accessor.global_accessor.model.CopyFrom(model_with_shapes)
+
+        # Build the graph outputs
+        graph_utils.build_graph_outputs(accessor.global_accessor.model, output)
+
+        # validate and check the model
+        onnx.checker.check_model(accessor.global_accessor.model, True)
+
+        return output
+
+
+class TrainingModel(building_blocks.Block):
+    """Builds the training model based on user's build method."""
+
+    def __init__(self):
+        super(TrainingModel, self).__init__()
+        self._arg_requiring_grad = set()
+        self._arg_not_requiring_grad = set()
+        self._parameters = None
+
+    @abstractmethod
+    def build(self, *args, **kwargs):
+        """Customize and build the forward graph for this training model.
+
+        This method is to be overridden by the user's implementation.
+        """
+        ...
+
+    def requires_grad(self, argument_name, value=True):
```

Comment:
Helpful to type annotate all newly introduced functions especially if they are public APIs.

Reply:
I will try to add annotations for the publicly exposed functions.